File: java-library/src/main/java/com/gblninja/parse/ParseTagType.java
package com.gblninja.parse;

import com.gblninja.tag.DefaultTag;
import com.gblninja.tag.GblType;
import com.gblninja.tag.Tag;
import com.gblninja.tag.TagHeader;
import com.gblninja.tag.type.GblBootloader;
import com.gblninja.tag.type.GblEnd;
import com.gblninja.tag.type.GblEraseProg;
import com.gblninja.tag.type.GblHeader;
import com.gblninja.tag.type.GblMetadata;
import com.gblninja.tag.type.GblProg;
import com.gblninja.tag.type.GblProgLz4;
import com.gblninja.tag.type.GblProgLzma;
import com.gblninja.tag.type.GblSeUpgrade;
import com.gblninja.tag.type.application.ApplicationData;
import com.gblninja.tag.type.application.GblApplication;
import com.gblninja.tag.type.certificate.ApplicationCertificate;
import com.gblninja.tag.type.certificate.GblCertificateEcdsaP256;
import com.gblninja.tag.type.certificate.GblSignatureEcdsaP256;
import com.gblninja.tag.type.encryption.GblEncryptionData;
import com.gblninja.tag.type.encryption.GblEncryptionInitAesCcm;
import com.gblninja.utils.ByteUtils;

public class ParseTagType {
    public static Tag parseTagType(long tagId, long length, byte[] byteArray) {
        GblType tagType = GblType.fromValue(tagId);
        TagHeader tagHeader = new TagHeader(tagId, length);

        if (tagType == null) {
            return new DefaultTag(
                    ByteUtils.copyArray(byteArray),
                    tagHeader,
                    GblType.TAG
            );
        }

        switch (tagType) {
            case HEADER_V3:
                long version = Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 0, 4).getInt());
                long gblType = Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 4, 4).getInt());
                return new GblHeader(tagHeader, tagType, version, gblType, byteArray);

            case BOOTLOADER:
                return new GblBootloader(
                        tagHeader,
                        tagType,
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 0, 4).getInt()),
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 4, 4).getInt()),
                        copyRange(byteArray, 8, byteArray.length),
                        byteArray
                );

            case APPLICATION:
                ApplicationData appData = new ApplicationData(
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 0, 4).getInt()),
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 4, 4).getInt()),
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 8, 4).getInt()),
                        (short) (byteArray[12] & 0xFF)
                );
                return new GblApplication(tagHeader, tagType, appData, byteArray);

            case METADATA:
                return new GblMetadata(tagHeader, tagType, byteArray, byteArray);

            case PROG:
                return new GblProg(
                        tagHeader,
                        tagType,
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 0, 4).getInt()),
                        copyRange(byteArray, 4, byteArray.length),
                        byteArray
                );

            case PROG_LZ4:
                return new GblProgLz4(tagHeader, tagType, byteArray);

            case PROG_LZMA:
                return new GblProgLzma(tagHeader, tagType, byteArray);

            case ERASEPROG:
                return new GblEraseProg(tagHeader, tagType, byteArray);

            case SE_UPGRADE:
                return new GblSeUpgrade(
                        tagHeader,
                        tagType,
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 0, 4).getInt()),
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 4, 4).getInt()),
                        copyRange(byteArray, 8, byteArray.length),
                        byteArray
                );

            case END:
                return new GblEnd(
                        tagHeader,
                        tagType,
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 0, 4).getInt()),
                        byteArray
                );

            case ENCRYPTION_DATA:
                return new GblEncryptionData(
                        tagHeader,
                        tagType,
                        copyRange(byteArray, 8, byteArray.length),
                        byteArray
                );

            case ENCRYPTION_INIT:
                return new GblEncryptionInitAesCcm(
                        tagHeader,
                        tagType,
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 0, 4).getInt()),
                        (short) (byteArray[4] & 0xFF),
                        byteArray
                );

            case SIGNATURE_ECDSA_P256:
                return new GblSignatureEcdsaP256(
                        tagHeader,
                        tagType,
                        byteArray,
                        (short) (byteArray[0] & 0xFF),
                        (short) (byteArray[1] & 0xFF)
                );

            case CERTIFICATE_ECDSA_P256:
                ApplicationCertificate certificate = new ApplicationCertificate(
                        (short) (byteArray[0] & 0xFF),
                        (short) (byteArray[1] & 0xFF),
                        (short) (byteArray[2] & 0xFF),
                        Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, 3, 4).getInt()),
                        (short) (byteArray[7] & 0xFF)
                );
                return new GblCertificateEcdsaP256(tagHeader, tagType, byteArray, certificate);

            default:
                return new DefaultTag(ByteUtils.copyArray(byteArray), tagHeader, GblType.TAG);
        }
    }

    private static byte[] copyRange(byte[] array, int start, int end) {
        if (start >= end || start >= array.length) return new byte[0];
        int actualEnd = Math.min(end, array.length);
        byte[] result = new byte[actualEnd - start];
        System.arraycopy(array, start, result, 0, actualEnd - start);
        return result;
    }
}

File: gbl/src/main/java/com/onix/gbl/parser/parse/ParseTagType.kt
package com.onix.gbl.parser.parse

import com.onix.gbl.parser.application.ApplicationData
import parser.data.tag.Tag
import parser.data.tag.TagHeader
import parser.data.tag.TagInterface
import parser.data.type.GblApplication
import parser.data.type.GblBootloader
import parser.data.type.GblEnd
import parser.data.type.GblEraseProg
import parser.data.type.GblHeader
import parser.data.type.GblMetadata
import parser.data.type.GblProg
import parser.data.type.GblProgLz4
import parser.data.type.GblProgLzma
import parser.data.type.GblSeUpgrade
import parser.data.type.GblType
import parser.data.utils.getFromBytes

fun parseTagType(
    tagId: UInt,
    length: UInt,
    byteArray: ByteArray,
): TagInterface {
    val tagType = GblType.fromValue(tagId.toLong())

    val tagHeader = TagHeader(
        id = tagId,
        length = length
    )

    return when (tagType) {
        GblType.HEADER_V3 -> {
            val version = getFromBytes(byteArray, offset = 0, length = 4).int.toUInt()
            val gblType = getFromBytes(byteArray, offset = 4, length = 4).int.toUInt()

            GblHeader(
                tagHeader = tagHeader,
                tagType = tagType,
                version = version,
                gblType = gblType,
                tagData = byteArray,
            )
        }

        GblType.BOOTLOADER -> {
            GblBootloader(
                tagHeader = tagHeader,
                tagType = tagType,
                bootloaderVersion = getFromBytes(byteArray, offset = 0, length = 4).int.toUInt(),
                address = getFromBytes(byteArray, offset = 4, length = 4).int.toUInt(),
                data = byteArray.copyOfRange(8, byteArray.size),
                tagData = byteArray
            )
        }

        GblType.APPLICATION -> {
            val appData = ApplicationData(
                type = getFromBytes(byteArray, offset = 0, length = 4).int.toUInt(),
                version = getFromBytes(byteArray, offset = 4, length = 4).int.toUInt(),
                capabilities = getFromBytes(byteArray, offset = 8, length = 4).int.toUInt(),
                productId = byteArray[12].toUByte(),
            )

            GblApplication(
                tagHeader = tagHeader,
                tagType = tagType,
                applicationData = appData,
                tagData = byteArray,
            )
        }

        GblType.METADATA -> {
            GblMetadata(
                tagHeader = tagHeader,
                tagType = tagType,
                metaData = byteArray,
                tagData = byteArray,
            )
        }

        GblType.PROG -> {
            GblProg(
                tagHeader = tagHeader,
                tagType = tagType,
                flashStartAddress = getFromBytes(byteArray, offset = 0, length = 4).int.toUInt(),
                data = byteArray.copyOfRange(4, byteArray.size),
                tagData = byteArray,
            )
        }

        GblType.PROG_LZ4 -> {
            GblProgLz4(
                tagHeader = tagHeader,
                tagType = tagType,
                tagData = byteArray,
            )
        }

        GblType.PROG_LZMA -> {
            GblProgLzma(
                tagHeader = tagHeader,
                tagType = tagType,
                tagData = byteArray,
            )
        }

        GblType.ERASEPROG -> {
            GblEraseProg(
                tagHeader = tagHeader,
                tagType = tagType,
                tagData = byteArray,
            )
        }

        GblType.SE_UPGRADE -> {
            GblSeUpgrade(
                tagHeader = tagHeader,
                tagType = tagType,
                blobSize = getFromBytes(byteArray, offset = 0, length = 4).int.toUInt(),
                version = getFromBytes(byteArray, offset = 4, length = 4).int.toUInt(),
                data = byteArray.copyOfRange(8, byteArray.size),
                tagData = byteArray,
            )
        }

        GblType.END -> {
            GblEnd(
                tagHeader = tagHeader,
                tagType = tagType,
                gblCrc = getFromBytes(byteArray, offset = 0, length = 4).int.toUInt(),
                tagData = byteArray,
            )
        }

        null -> {
            Tag(
                tagType = GblType.TAG,
                tagHeader = tagHeader,
                tagData = byteArray
            )
        }

        else -> {
            Tag(
                tagType = GblType.TAG,
                tagHeader = tagHeader,
                tagData = byteArray
            )
        }
    }
}